[Polygon2DEditor] Add tool to automatically move center of gravity to origin.

bruvzg opened the pull request:
Related to https://github.com/godotengine/godot/issues/103965, can be used to fix similar problems or center messy polygon points in general.

https://github.com/user-attachments/assets/751fab5b-586c-471b-bc16-001421f472fd

Edit: added origin and center point display.

bruvzg committed:
[Polygon2DEditor] Add tool to automatically move center of gravity to origin.

groud commented:
I think the feature is nice, but it would be even better to have the center of mass visible (and, even better, movable).

bruvzg commented:
> but it would be even better to have the center of mass/origin visible (and, even better, movable).

Done, both origin and center of mass are shown, and center is movable.

groud commented:
> > but it would be even better to have the center of mass/origin visible (and, even better, movable).
> 
> Done, both origin and center of mass are shown, and center is movable.

I am not sure I follow. Wasn't the origin considered as the center of mass ? I think it's fine if they are kept as the same point, I just think it should be visible. I feel this is a lot of code added to the abstract_polygon editor for something that sounds a bit specific too, but maybe it can be useful elsewhere I guess.

bruvzg commented:
> Wasn't the origin considered as the center of mass

For physics, yes. By "and center is movable" I meant moving the polygon using its geometric center.

- Tool button moves origin to the center of the polygon.

<img width="829" alt="Screenshot 2025-03-14 at 16 14 24" src="https://github.com/user-attachments/assets/fb144b6c-5b6c-4513-994d-a07e018075b0" />

- Center point can be freely dragged, moving the whole polygon relative to the origin.

https://github.com/user-attachments/assets/e7958acb-7cf0-4ea6-b96b-d97a08f720ba

Since origin is also center of rotation/transform, it's not always desired to have it and center of polygon in the same place. And there was no way to move it without moving all points individually.

groud commented:
Ah ok, I get it. The geometric center is just used as an handle then. Why not simply move the polygon when dragging in the red area then ? We could keep the current behaviors when close the the edges, but move the polygon instead if we are far from them?
I find it a bit weird to use a single point to move a whole polygon tbh, though I guess it works fine.

JekSun97 commented:
Let's add this!

Calinou commented on editor/plugins/abstract_polygon_2d_editor.cpp:
```diff
@@ -513,11 +638,34 @@ void AbstractPolygon2DEditor::forward_canvas_draw_over_viewport(Control *p_overl
+
+	Ref<Font> font = get_theme_font(SNAME("bold"), EditorStringName(EditorFonts));
+	int font_size = 1.3 * get_theme_font_size(SNAME("bold_size"), EditorStringName(EditorFonts));
+	const float outline_size = 4;
```
The outline size should be multiplied by `EDSCALE`.

Calinou reviewed (commented):
Tested locally, it works as expected. This is a very welcome usability improvement :slightly_smiling_face: 

One potential issue I noticed is that the default marker icon can be seen behind the `O` icon. This is most noticeable at high editor scales:

![image](https://github.com/user-attachments/assets/232b803a-dd3d-4db7-8eea-2e73957a710d)

I'm not sure if there's a way to make it invisible without affecting interactivity (maybe setting an empty icon with the correct size works).

KoBeWi commented:
The button says "Move center of gravity to origin.", but it appears to be doing the opposite, i.e. it moves origin point onto the center.

Also is there a way to somehow explain the meaning of the two points? I imagine they'll look cryptic for unfamiliar users and the button's tooltip is the only clue.

KoBeWi commented on editor/plugins/abstract_polygon_2d_editor.cpp:
```diff
@@ -128,6 +128,43 @@ bool AbstractPolygon2DEditor::_has_resource() const {
+	cx /= n_subs;
+	cy /= n_subs;
+
+	return (Vector2(cx, cy));
```
```suggestion
	return Vector2(cx, cy);
```
Too many brackets.

KoBeWi commented on editor/plugins/abstract_polygon_2d_editor.cpp:
```diff
@@ -238,15 +305,18 @@ void AbstractPolygon2DEditor::disable_polygon_editing(bool p_disable, const Stri
 		button_create->set_tooltip_text(TTR("Create points."));
 		button_edit->set_tooltip_text(TTR("Edit points.\nLMB: Move Point\nRMB: Erase Point"));
 		button_delete->set_tooltip_text(TTR("Erase points."));
+		button_center->set_tooltip_text(TTR("Move center of gravity to origin."));
```
```suggestion
		button_center->set_tooltip_text(TTRC("Move center of gravity to origin."));
```
All of the buttons should use TTRC, but you don't need to update them.

KoBeWi commented on editor/plugins/abstract_polygon_2d_editor.cpp:
```diff
@@ -513,11 +638,34 @@ void AbstractPolygon2DEditor::forward_canvas_draw_over_viewport(Control *p_overl
 	const bool is_closed = !_is_line();
 
+	const Vector2 &center = _get_geometric_center();
+	if (!center.is_equal_approx(Vector2())) {
```
```suggestion
	if (!center.is_zero_approx()) {
```

bruvzg commented:
> The origin and center point also appear for Polygon2D. I think they shouldn't? Center of mass is irrelevant for Polygon2D, and there is also the Polygon bottom editor that allows moving points.

Yes, It's only relevant for collisions shapes. Limited it to collision shapes.

> When you right-click while dragging center, it will open 2D editor's menu and make the dragging stuck. It does not happen when dragging regular points.

Fixed.

> One potential issue I noticed is that the default marker icon can be seen behind the O icon. This is most noticeable at high editor scales

Both are non-interactive and always at the same place, so not sure what's the issue. Default maker also under the polygon itself and in most cases not visible. But since I have limited it to collision polygons, and collision polygons does not have default marker, it's no longer relevant.

> The button says "Move center of gravity to origin.", but it appears to be doing the opposite, i.e. it moves origin point onto the center.

Renamed it to `Move center of gravity to geometric center` to avoid confusion. Physical center of gravity is origin (“O” point), and “C” point is geometric center.
